grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImplIterator.java

```diff
@@ -41,43 +43,43 @@
  */
 class ReasonerQueryImplIterator extends ReasonerQueryIterator {
 
-    private final Answer partialSubstitution;
+    private Answer partialSub = new QueryAnswer();
     private final ReasonerQueryImpl queryPrime;
 
     private final QueryCache<ReasonerAtomicQuery> cache;
     private final Set<ReasonerAtomicQuery> subGoals;
 
-    private Iterator<Answer> queryIterator;
     private final Iterator<Answer> atomicQueryIterator;
+    private Iterator<Answer> queryIterator = Collections.emptyIterator();
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReasonerQueryImpl.class);
 
-    ReasonerQueryImplIterator(ReasonerQueryImpl query,
+    ReasonerQueryImplIterator(ReasonerQueryImpl q,
                               Answer sub,
                               Set<ReasonerAtomicQuery> subGoals,
                               QueryCache<ReasonerAtomicQuery> cache){
-        this.partialSubstitution = sub;
         this.subGoals = subGoals;
         this.cache = cache;
 
         //get prioritised atom and construct atomic query from it
-        this.queryPrime = new ReasonerQueryImpl(query);
-        queryPrime.addSubstitution(sub);
-        Atom topAtom = queryPrime.getTopAtom();
-        ReasonerAtomicQuery q = new ReasonerAtomicQuery(topAtom);
+        ReasonerQueryImpl query = new ReasonerQueryImpl(q);
+        query.addSubstitution(sub);
+        Atom topAtom = query.getTopAtom();
 
-        boolean isAtomic = queryPrime.isAtomic();
-        if (!isAtomic) queryPrime.removeAtom(topAtom);
+        LOG.trace("CQ: " + query);
+        LOG.trace("CQ delta: " + sub);
 
-        atomicQueryIterator = isAtomic? Collections.emptyIterator() : q.iterator(subGoals, cache);
-        queryIterator = isAtomic? q.iterator(subGoals, cache) : Collections.emptyIterator();
+        this.atomicQueryIterator = new ReasonerAtomicQuery(topAtom).iterator(new QueryAnswer(), subGoals, cache);
+        this.queryPrime = ReasonerQueries.prime(query, topAtom);
     }
 
     @Override
     public boolean hasNext() {
         if (queryIterator.hasNext()) return true;
         else {
```

Comment:
Once again the `else` is not needed. Check all the implementations of `ReasonerQueryIterator#hasNext` you may have the possibility of getting rid of a lot of boilerplate stuff.